File: Editorial/89_Gray_Code/Python3/4_Recursion_2.py
class Solution:
    def __init__(self):
        self.nextNum = 0

    def grayCode(self, n):
        self.result = []
        self.grayCodeHelper(n)
        return self.result

    def grayCodeHelper(self, n):
        if n == 0:
            self.result.append(self.nextNum)
            return
        self.grayCodeHelper(n - 1)
        # Flip the bit at (n - 1)th position from right
        self.nextNum = self.nextNum ^ (1 << (n - 1))
        self.grayCodeHelper(n - 1)

File: Editorial/89_Gray_Code/CSharp/4_Recursion_2.cs
public class Solution {
    private int nextNum = 0;

    public IList<int> GrayCode(int n) {
        List<int> result = new List<int>();
        GrayCodeHelper(result, n);
        return result;
    }

    private void GrayCodeHelper(List<int> result, int n) {
        if (n == 0) {
            result.Add(this.nextNum);
            return;
        }

        GrayCodeHelper(result, n - 1);
        // Flip the bit at (n - 1)th position from right
        this.nextNum = this.nextNum ^ (1 << (n - 1));
        GrayCodeHelper(result, n - 1);
    }
}
